native/cocos/renderer/pipeline/shadow/CSMLayers.cpp

```diff
@@ -0,0 +1,298 @@
+        _matShadowProj = matShadowProj.clone();
+        _matShadowViewProj = matShadowViewProj.clone();
+    }
+    _validFrustum.createOrtho(orthoSizeWidth, orthoSizeHeight, 0.1F, _shadowCameraFar, matShadowTrans);
+}
+
+void ShadowTransformInfo::specialLayerCopyValidFrustum(const geometry::Frustum &validFrustum) {
+    geometry::Frustum::copy(&_validFrustum, validFrustum);
+}
+
+void ShadowTransformInfo::specialLayerCreateOrtho(float width, float height, float nearClamp, float farClamp, const Mat4 &transform) {
+    _validFrustum.createOrtho(width, height, nearClamp, farClamp, transform);
+}
+
+void ShadowTransformInfo::specialLayerSplitFrustum(float start, float end, float aspect, float fov, const Mat4 &transform) {
+    _splitFrustum.split(start, end, aspect, fov, transform);
+}
+
+void ShadowTransformInfo::splitFrustum(float start, float end, float aspect, float fov, const Mat4& transform) {
+    _splitFrustum.split(start, end, aspect, fov, transform);
+}
+
+CSMLayerInfo::CSMLayerInfo(uint level) : ShadowTransformInfo(level) {
+    _matShadowAtlas.setZero();
+    _matShadowViewProjAtlas.setZero();
+    calculateAtlas(level);
+}
+
+void CSMLayerInfo::calculateAtlas(uint level) {
+    const gfx::Device *device = gfx::Device::getInstance();
+    const float clipSpaceSignY = device->getCapabilities().clipSpaceSignY;
+    const float x = floorf(static_cast<float>(level % 2U)) - 0.5F;
+    const float y = clipSpaceSignY * (0.5F - floorf(static_cast<float>(level) / 2U));
+    const Vec3 bias(x, y, 0.0F);
+    const Vec3 scale(0.5F, 0.5F, 1.0F);
+    Mat4::fromRTS(Quaternion::identity(), bias, scale, &_matShadowAtlas);
+}
+
+CSMLayers::CSMLayers() {
+    _specialLayer = ccnew ShadowTransformInfo(1U);
+    _layers.reserve(static_cast<size_t>(scene::CSMLevel::LEVEL_4));
+}
+
+CSMLayers::~CSMLayers() {
+    CC_SAFE_DELETE(_specialLayer);
+
+    for (const auto *shadowCSMLayer : _layers) {
+        CC_SAFE_DELETE(shadowCSMLayer);
+    }
+}
+
+void CSMLayers::update(const PipelineSceneData *sceneData, const scene::Camera *camera) {
+    CC_ASSERT(sceneData);
+    CC_ASSERT(camera);
+
+    const scene::Shadows *shadowInfo = sceneData->getShadows();
+    const scene::RenderScene *const scene = camera->getScene();
+    scene::DirectionalLight *dirLight = scene->getMainLight();
+
+    CC_ASSERT(dirLight);
+
+    const uint levelCount = static_cast<uint>(dirLight->getShadowCSMLevel());
+    CC_ASSERT(levelCount <= static_cast<uint>(scene::CSMLevel::LEVEL_4));
+    const float shadowDistance = dirLight->getShadowDistance();
+
+    if (!shadowInfo->isEnabled() || !dirLight->isShadowEnabled()) {
+        return;
+    }
+
+    if (dirLight->isShadowFixedArea()) {
+        updateFixedArea(dirLight);
+    } else {
+        size_t oldLayerSize = _layers.size();
+        const bool isResized = oldLayerSize < levelCount;
+        while (oldLayerSize < levelCount) {
+            _layers.emplace_back(ccnew CSMLayerInfo(static_cast<uint>(oldLayerSize)));
```

Comment:
emplace_back will append a element at the end of the vector. As _layers has fixed length, it is better to use `std::array` instead.